### PR TITLE
AK: pow() with less UB and more performance

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
(1) remove UB by only taking the integer fast path if the exponent fits in the i64 integer limits.

(2) use exponentiation by squaring for the integer fast path, thereby actually making it fast (and not time out on the new tests...)

### CODE REVIEW

**Review by Hendiadyoin1 — APPROVED** [review]
Funfact: The integer approximation of powers is technically a bit inaccurate (to my knowledge)
Also: Why don't we use exponentiation by squaring here?

File: AK/Math.h [hunk 1]
```diff
@@ -1013,14 +1013,18 @@ constexpr T pow(T x, T y)
...
-    int y_as_int = (int)y;
-    if (y == (T)y_as_int) {
-        T result = x;
-        for (int i = 0; i < fabs<T>(y) - 1; ++i)
-            result *= x;
-        if (y < 0)
-            result = 1.0l / result;
-        return result;
+
+    // Take an integer fast path as long as the value fits within a 64-bit integer.
```
Comment (gmta) [inline AK/Math.h:1017]: Can you expose this issue with a test?

File: Tests/AK/TestAKMath.cpp [hunk 2]
```diff
@@ -0,0 +1,2305 @@
...
+#include <AK/Math.h>
+
+struct MathTest {
+    double input;
+    double expected_output;
+};
+
+// Generated with Python + numpy:
+// echo -e "import numpy as np\nfor x in np.arange(-1020, 1020, 0.9):\n    print(f'{{ {x}, {2**x} }},')" | python3
+// FIXME: Actually use these for a test of AK::pow; currently many inputs are broken due to our bad implementation.
```
Comment (gmta) [inline Tests/AK/TestAKMath.cpp:19]: Let's not add something that's essentially a 2200+ line FIXME. If you insist on having a FIXME, just leave the intent and the code to generate the table please.

File: Tests/AK/TestAKMath.cpp [hunk 3]
```diff
@@ -0,0 +1,2305 @@
...
+    EXPECT_EQ(AK::pow(10., 4.), 10'000.);
+    EXPECT_EQ(AK::pow(10., 5.), 100'000.);
+    EXPECT_EQ(AK::pow(10., 6.), 1'000'000.);
+    EXPECT_EQ(AK::pow(10., 19.), 10'000'000'000'000'000'000.);
+    // Should not take the integer fast path anymore, since we're beyond 64-bit integers.
+    EXPECT_EQ(AK::pow(10., 20.), 100'000'000'000'000'000'000.);
+    EXPECT_EQ(AK::pow(10., 21.), 1'000'000'000'000'000'000'000.);
+    EXPECT_EQ(AK::pow(10., 22.), 10'000'000'000'000'000'000'000.);
+    EXPECT_EQ(AK::pow(10., 23.), 100'000'000'000'000'000'000'000.);
+}
```
Comment (gmta) [inline Tests/AK/TestAKMath.cpp:2305]: So when I run this test case on `master`, it succeeds. Can you add a test that fails on `master` and succeeds after your PR?

Comment (kleinesfilmroellchen) [reply Tests/AK/TestAKMath.cpp:2305]: Huh. That's clearly UB though.

Comment (kleinesfilmroellchen) [reply Tests/AK/TestAKMath.cpp:2305]: My test case was in Maps by the way, so I don't have numbers unfortunately

**Comment by kleinesfilmroellchen** [comment]
@gmta The "problem" with these tests is that they yield correct results since the integer fast path is bailed out of with the comparison y == int_y, but the UB occurs before that. So this is really a test for triggering UB; it passes on master without UBSan just fine.

File: Tests/AK/CMakeLists.txt [hunk 4]
```diff
@@ -48,6 +48,7 @@ set(AK_TEST_SOURCES
     TestLEB128.cpp
     TestLexicalPath.cpp
     TestMACAddress.cpp
+    TestAKMath.cpp
```
Comment (gmta) [inline Tests/AK/CMakeLists.txt:51]: Let's put this at the top so it's in alphabetical order

**Comment by gmta** [comment]
> So this is really a test for triggering UB; it passes on master without UBSan just fine.

Gotcha, seems fine to deal with the UB this way.

File: Tests/AK/TestAKMath.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    EXPECT_EQ(AK::pow(10., 0.), 1.);
+    EXPECT_EQ(AK::pow(10., 1.), 10.);
+    EXPECT_EQ(AK::pow(10., 2.), 100.);
+    EXPECT_EQ(AK::pow(10., 3.), 1'000.);
+    EXPECT_EQ(AK::pow(10., 4.), 10'000.);
+    EXPECT_EQ(AK::pow(10., 5.), 100'000.);
+    EXPECT_EQ(AK::pow(10., 6.), 1'000'000.);
+    EXPECT_EQ(AK::pow(1.e18, 2.), 1.e36);
+    EXPECT_EQ(AK::pow(1.e19, 2.), 1.e38);
+    // Should not take the integer fast path anymore, since we're beyond 64-bit integers.
```
Comment (gmta) [inline Tests/AK/TestAKMath.cpp:32]: Your change in `AK::pow` relates to the `y` argument, which is `2.` in these tests. How would that exceed a 64-bit integer?

Comment (kleinesfilmroellchen) [reply Tests/AK/TestAKMath.cpp:32]: The result exceeds 64 bits, not the arguments.

Comment (gmta) [reply Tests/AK/TestAKMath.cpp:32]: Your comment states "should not take the integer fast path", but that has nothing to do with the result being beyond a 64-bit integer. The fast path seems to be dependent on the `y` argument being equal to itself casted to an `i64` and fitting withing the `i64` limits.

Comment (gmta) [reply Tests/AK/TestAKMath.cpp:32]: Just to be completely clear, these are the conditions for the fast path:

```cpp
    if (y >= static_cast<T>(NumericLimits<i64>::min()) && y <= static_cast<T>(NumericLimits<i64>::max())) {
        i64 y_as_int = static_cast<i64>(y);
        if (y == static_cast<T>(y_as_int)) {
```

So having `y == 2.` does not test the absence of a fast path afaics.

**Comment by gmta** [comment]
CI failure.

**Comment by nico** [comment]
What's with the bots here? Looks like something real, somehow?

**Comment by kleinesfilmroellchen** [comment]
> What's with the bots here? Looks like something real, somehow?

It indeed is something real. I'm fairly confident that the cast from the 64-bit min/max values to double round in the wrong direction, so the safeguard against the fast path doesn't work when we're just around the integer limit.

**Comment by kleinesfilmroellchen** [comment]
Possibly fixed with some adjusted check bounds, let's see if CI agrees with my local UBSan.

File: AK/Math.h [hunk 6]
```diff
@@ -996,14 +997,19 @@ constexpr T pow(T x, T y)
...
+    // To avoid an incorrect comparison due to rounding, we slightly adjust the 64-bit bounds towards zero.
+    if (y >= nextafter(static_cast<T>(NumericLimits<i64>::min()), Infinity<T>) && y <= nextafter(static_cast<T>(NumericLimits<i64>::max()), -Infinity<T>)) {
```
Comment (BertalanD) [inline AK/Math.h:1003]: What in the world? This blog post shows a similar logic working correctly without `nextafter` magic:

https://sillycross.github.io/2022/07/18/2022-07-18/#Pick-the-Right-Tool-Instead

Comment (kleinesfilmroellchen) [reply AK/Math.h:1003]: Well, it clearly fixed the UB, so I do not know. I would have preferred the normal solution too.

Comment (kleinesfilmroellchen) [reply AK/Math.h:1003]: That blog's magic values are what i am "synthesizing" here though...

Comment (kleinesfilmroellchen) [reply AK/Math.h:1003]: from earlier in the post

> While -2^63 (the smallest int64_t) has an exact representation in double, 2^63-1 (the largest int64_t) doesn’t. So we must be careful about the rounding problems when doing the comparison.

Comment (kleinesfilmroellchen) [reply AK/Math.h:1003]: Re-checked this, and if I change the lower bound to just the 64-bit min, our formatter says -9223372036854776000, which should actually be -9223372036854775808, which is the same as in the blogpost. For the upper bound, we have a value that is slightly off, but I my error was to use <= instead of <. With that, it should work with just static_cast'ing the min and max values, which I think is more readable than these huge constants.

**Comment by kleinesfilmroellchen** [comment]
I'm not sure what those infinite loops are doing, possibly related to the naive exponentiation code? I can't reproduce that locallly. If the issue persists I will just add another commit with exponentiation by squaring.

**Comment by kleinesfilmroellchen** [comment]
CI seems flaky; apt.llvm.org is down?

File: AK/Math.h [hunk 7]
```diff
@@ -984,11 +984,25 @@ using Hyperbolic::cosh;
 using Hyperbolic::sinh;
 using Hyperbolic::tanh;
 
+// Calculate x^y with fast exponentiation when the power is an integer.
+template<FloatingPoint F, Integral I>
+constexpr F pow_int(F x, I y)
```
Comment (Hendiadyoin1) [inline AK/Math.h:989]: As this only works for Unsigned y, maybe we could enforce that

Comment (kleinesfilmroellchen) [reply AK/Math.h:989]: Yeah, immediately noticed after pushing that that was wrong. It won’t break at the moment since we check against i64 bounds, which are smaller than u64 bounds in the positive.

File: Tests/AK/TestAKMath.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    EXPECT_EQ(AK::pow(1.e18, 2.), 1.e36);
+    EXPECT_EQ(AK::pow(1.e19, 2.), 1.e38);
+    EXPECT_EQ(AK::pow(1.e20, 2.), 1.e40);
+    EXPECT_EQ(AK::pow(1.e21, 2.), 1.e42);
+    EXPECT_EQ(AK::pow(1.e22, 2.), 1.e44);
+    // Largest 64-bit float value that fits in a 64-bit integer.
+    EXPECT_EQ(AK::pow(1., 9223372036854774784.0), 1.);
+    // Should not take the integer fast path anymore, since the exponent is beyond 64-bit integers.
+    EXPECT_EQ(AK::pow(1., 9223372036854775808.0), 1.);
+    EXPECT_EQ(AK::pow(1., 9223372036854775810.0), 1.);
```
Comment (DanShaders) [inline Tests/AK/TestAKMath.cpp:33]: `9223372036854775810. == 9223372036854775808.`.

The next actually different double is 9223372036854777856.

File: Tests/AK/TestAKMath.cpp [hunk 9]
```diff
@@ -30,5 +30,5 @@ TEST_CASE(pow)
     EXPECT_EQ(AK::pow(1., 9223372036854774784.0), 1.);
     // Should not take the integer fast path anymore, since the exponent is beyond 64-bit integers.
     EXPECT_EQ(AK::pow(1., 9223372036854775808.0), 1.);
-    EXPECT_EQ(AK::pow(1., 9223372036854775810.0), 1.);
+    EXPECT_EQ(AK::pow(1., 9223372036854777856.0), 1.);
```
Comment (nico) [inline Tests/AK/TestAKMath.cpp:33]: Why this change in this commit?

File: AK/Math.h [hunk 10]
```diff
@@ -996,14 +997,18 @@ constexpr T pow(T x, T y)
...
-    if (y == (T)y_as_int) {
-        T result = x;
-        for (int i = 0; i < fabs<T>(y) - 1; ++i)
-            result *= x;
-        if (y < 0)
-            result = 1.0l / result;
-        return result;
+
+    // Take an integer fast path as long as the value fits within a 64-bit integer.
+    if (y >= static_cast<T>(NumericLimits<i64>::min()) && y < static_cast<T>(NumericLimits<i64>::max())) {
```
Comment (nico) [inline AK/Math.h:1016]: Shouldn't this be `<= NumericLimits<i64>::max()` on the rhs?

Comment (kleinesfilmroellchen) [reply AK/Math.h:1016]: Nope, that caused issues before.

File: Tests/AK/TestAKMath.cpp (hunk 9 above)
Comment (kleinesfilmroellchen) [reply Tests/AK/TestAKMath.cpp:33]: because rebasing old commits is hard, ugh

File: AK/Math.h (hunk 10 above)
Comment (nico) [reply AK/Math.h:1016]: But we use `static_cast<T>(NumericLimits<i64>::min())` on the lhs, and fabs() of that is larger than::max().

What kind of issues, and why don't they happen with the negative limit?

Comment (kleinesfilmroellchen) [reply AK/Math.h]: okay okay i will let CI decide, please let me be done with this commit already

**Comment by kleinesfilmroellchen** [comment]
@nico I hope you can see why I did the `<` check for the upper limit, since it prevents the conversion overflow properly. (The CI tests ran fine before that change.) I would now like to change this back to the version that works and get this merged, okay?

File: AK/Math.h (hunk 6 above)
Comment (DanShaders) [reply AK/Math.h]: > our formatter says -9223372036854776000, which should actually be -9223372036854775808

And our formatter is completely correct because the value it outputs is precisely the shortest roundtrippable decimal representation for this double.

File: AK/Math.h [hunk 11]
```diff
@@ -1002,15 +1016,14 @@ constexpr T pow(T x, T y)
     if (y >= static_cast<T>(NumericLimits<i64>::min()) && y < static_cast<T>(NumericLimits<i64>::max())) {
         i64 y_as_int = static_cast<i64>(y);
         if (y == static_cast<T>(y_as_int)) {
-            T result = x;
-            for (u64 i = 0; i < static_cast<u64>(fabs<T>(y) - 1); ++i)
-                result *= x;
+            T result = pow_int(x, static_cast<u64>(fabs<T>(y)));
             if (y < 0)
                 result = 1.0l / result;
```
Comment (Hendiadyoin1) [inline AK/Math.h:1021]: Nit, force this 1 to be of type T while you are here

**Review by timschumi — APPROVED** [review]
